### ISSUE TITLE
Labels are not clickable anymore

### ISSUE THREAD

**Issue by stianjensen** [philc/vimium#3492]
**Describe the bug**
When pressing 'f' to highlight click targets, labels used to be highlighted so that you were able to select for instance custom bootstrap checkboxes (https://getbootstrap.com/docs/4.4/components/forms/#checkboxes)
Since the last version, those checkboxes are no longer selectable with vimium.

**To Reproduce**
1. Go to URL: https://getbootstrap.com/docs/4.4/components/forms/#checkboxes
2. Click 'f' on the keyboard
3. Observe that the checkbox does not get highlighted

Before and after:
<pre align=center><img height=350 alt src="https://user-images.githubusercontent.com/1413267/74238723-a305cc80-4cd6-11ea-91b1-2c7bd92b3c31.png"><img height=350 alt src="https://user-images.githubusercontent.com/1413267/74238725-a5682680-4cd6-11ea-826e-5c863ee74be3.png"></pre>

**Browser and Vimium version**
I've seen the bug in both Chrome and Firefox, although I don't have a specific chrome version handy at the moment.

Firefox Version: 73.0b12
Build ID 	20200130174018
Vimium 1.65.0, 1.65.1, 1.65.2

(A working version of Vimium is 1.64.6)

**Comment by mihmig** [comment]
span, label and other elements, that have CSS-attribute cursor, might also be clickable:
![изображение](https://user-images.githubusercontent.com/4863752/74632007-71c24c00-516f-11ea-9c78-e5efc7546c28.png)

**Comment by lydell** [comment]
As far as I can tell, those checkboxes are actually standard `<input type="checkbox">` so the checkboxes themselves should be detectable by Vimium. @poacher2k You might be interested in this.

**Comment by poacher2k** [comment]
I’ll check it out! Thanks for the mention.

**Comment by poacher2k** [comment]
As far as I can tell, what's going on here is that the input have `psoition: absolute` and `z-index: -1`. No idea why the label doesn't have a link hint, but it didn't have that in the before picture either 🤔

**Comment by stianjensen** [comment]
Okay, so the change isn't that labels aren't clickable anymore, but rather that hidden checkboxes aren't clickable anymore.

I guess that ignoring hidden checkboxes (and other elements) has other advantages, so you don't simply want to revert that change (I see other webpages with hidden popup menus have improved in the newest version).

But maybe labels in general should get link hints - at least those with the `for` attribute set?

**Comment by poacher2k** [comment]
> But maybe labels in general should get link hints - at least those with the for attribute set?

Totally agree. Thoughts @philc ?

**Comment by lydell** [comment]
Sorry for the confusion – you’re right. The checkbox _is_ hidden in this case, and replaced with `label::before` and `label::after` styled as a checkbox.

I was confused becuase my own extension – [Link Hints](https://github.com/lydell/LinkHints/) – _does_ identify the checkbox as visible and places a hint near it (rather than falling back to adding a hint for the label instead), so I assumed that Vimium had some little bug in its `.elementFromPoint` usage. But it turned out to be much trickier than that: https://github.com/lydell/LinkHints/commit/704f1d414eae28cce2604584bcf03f65fbb197fe

I suspect Vimium has _never_ supported adding hints for `<label>` elements. Vimium identifies `<input type="checkbox">` as clickable, but `.elementFromPoint()` returns the `<label>` element here, because its pseudo elements are located on top of the checkbox, making Vimium think the checkbox is covered by other elements and therefore shouldn’t get a hint. The reason this worked before is simply because the `.elementFromPoint()` check didn’t use to exist. You should be able to fix this by using the same workaround as in Link Hints: https://github.com/lydell/LinkHints/blob/704f1d414eae28cce2604584bcf03f65fbb197fe/src/worker/ElementManager.js#L2066-L2068

That being said, it would be good to support hints for `<label>` elements as well. The only tricky part is to _not_ give hints to `<label>`s if their form control is visible (otherwise you end up with basically duplicate hints for every form control which is a bit noisy and makes the hints longer). (Yes, Link Hints does this as well ^.^)

**Comment by dannyharding10** [comment]
Any update on this issue? I would love to be able to review a whole Merge Request on Gitlab with vimium but I can't collapse the diffs using the checkbox as I would hope to be able to 
![image](https://user-images.githubusercontent.com/11875246/194153998-ad39face-9fab-43b3-ab85-94e3f1b3a313.png)

**Comment by poacher2k** [comment]
> That being said, it would be good to support hints for `<label>` elements as well. The only tricky part is to _not_ give hints to `<label>`s if their form control is visible (otherwise you end up with basically duplicate hints for every form control which is a bit noisy and makes the hints longer). (Yes, Link Hints does this as well ^.^)

One option could be to _only_ give it for the label and not the input if the label is a parent of the input, or is referred to with `for="inputId"`. Not really sure what the best solution is though.

**Comment by dannyharding10** [comment]
For what it's worth, I would much prefer duplicate tags over no tags at all. As you can see in my screenshot there are already plenty of duplicates which is fine

**Comment by gdh1995** [comment]
@poacher2k Vimium has used a logic to show hint for a `<label>` only if its bound `<input>` is not "visible".

@dannyharding10 today I tested Vimium 1.67.1 on https://gitlab.com/gitlab-org/omnibus-gitlab/-/merge_requests/6383/diffs and it did show hints for "`viewed`" checkboxes, so I wonder what's the version of you tested GitLab website. Maybe it's too old to provide standard information about whether checkboxes are clickable or not.

**Comment by poacher2k** [comment]
> @poacher2k Vimium has used a logic to show hint for a `<label>` only if its bound `<input>` is not "visible".
> 
> @dannyharding10 today I tested Vimium 1.67.1 on https://gitlab.com/gitlab-org/omnibus-gitlab/-/merge_requests/6383/diffs and it did show hints for "`viewed`" checkboxes, so I wonder what's the version of you tested GitLab website. Maybe it's too old to provide standard information about whether checkboxes are clickable or not.

Sounds perfect! But I can confirm what @dannyharding10 experiences - the label isn't clickable on this page, at least not in Firefox:
<img width="356" alt="image" src="https://user-images.githubusercontent.com/2255960/194810583-c9b6cf8a-95c4-42bf-8001-02d3e109ce54.png">

The visible checkbox is a `:before` element, while the actual checkbox has `position: absolute; z-index: -1;`, meaning the actual checkbox is not visible. If I manually set `display: none` on the checkbox and press F, the label correctly gets a link hint.

**Comment by gdh1995** [comment]
Yes your analysis is correct, but it's strange that my Firefox 105 on Win11 does give different result:
![image](https://user-images.githubusercontent.com/5547703/194903959-2209f56a-5357-4253-a42e-b1dfa1071d30.png)
- The test link is https://gitlab.com/gitlab-org/omnibus-gitlab/-/merge_requests/6383/diffs , and screen DPI is 100% / 200%.
- `elementFromPoint(center)` returns the `label:before`
- while `elementFromPoint(left+0.1, top+0.1)` returns the `input`
  - I don't know why

You may try my customized version of Vimium, named Vimium C (https://github.com/gdh1995/vimium-c), and it has many improved details to try its best to find clickable elements.
